receiver/azuremonitorreceiver/scraper.go

```diff
@@ -319,6 +338,64 @@ func (s *azureScraper) getResources(ctx context.Context, subscriptionID string)
 	s.subscriptions[subscriptionID].resourcesUpdated = time.Now()
 }
 
+// hackResources is a workaround specially done for the storageAccount metrics.
+// Every StorageAccount resources have some implicit sub resources (/blobServices/default, fileServices/default, etc...) that are not returned by the API.
+// We need to add them manually to the resources and resourceTypes map.
+// Note that we do that hack only if the user has asked the sub resource types explicitly in the services config.
+// Example:
+// For each resource with id .../Microsoft.Storage/storageAccount/myResource of type Microsoft.Storage/storageAccount,
+// It will create a fake resource with id .../Microsoft.Storage/storageAccount/myResource/blobServices/default of type Microsoft.Storage/storageAccounts/blobServices.
+// TODO: duplicate
+func (s *azureScraper) hackResources(resources []*armresources.GenericResourceExpanded) []*armresources.GenericResourceExpanded {
+	var hackedResources []*armresources.GenericResourceExpanded
+	for _, resource := range resources {
+		hackedResources = append(hackedResources, resource)
+		if resource.Type != nil && *resource.Type == storageAccountType {
+			if s.storageAccountHackConfig.askedBlobServices {
+				if r := s.hackResource(resource, "Microsoft.Storage/storageAccounts/blobServices", fmt.Sprintf("%s/blobServices/default", *resource.ID)); r != nil {
+					hackedResources = append(hackedResources, r)
+				}
+			}
+			if s.storageAccountHackConfig.askedFileServices {
+				if r := s.hackResource(resource, "Microsoft.Storage/storageAccounts/fileServices", fmt.Sprintf("%s/fileServices/default", *resource.ID)); r != nil {
+					hackedResources = append(hackedResources, r)
+				}
+			}
+			if s.storageAccountHackConfig.askedQueueServices {
+				if r := s.hackResource(resource, "Microsoft.Storage/storageAccounts/queueServices", fmt.Sprintf("%s/queueServices/default", *resource.ID)); r != nil {
+					hackedResources = append(hackedResources, r)
+				}
+			}
+			if s.storageAccountHackConfig.askedTableServices {
+				if r := s.hackResource(resource, "Microsoft.Storage/storageAccounts/tableServices", fmt.Sprintf("%s/tableServices/default", *resource.ID)); r != nil {
+					hackedResources = append(hackedResources, r)
+				}
+			}
+		}
+	}
+	return hackedResources
+}
+
+// hackResource creates a fake new resource with give type and ID.
+// The rest of the attributes (location, tags, etc...) are copied from the original resource.
+// TODO: duplicate
+func (s *azureScraper) hackResource(resource *armresources.GenericResourceExpanded, newResourceType, newResourceID string) *armresources.GenericResourceExpanded {
+	hackedResource, err := copyResource(resource)
+	if err != nil {
+		s.settings.Logger.Error("failed to create a hack resource",
+			zap.String("resource type", *resource.Type),
```

Comment:
I believe there is a stringp type that does the check to see if the pointer is set or not.